### ISSUE TITLE
possible race condition in JsonFormatter?

### ISSUE THREAD

**Issue by dfa1** [quarkiverse/quarkus-logging-json#111]
In our application occasionally we have random characters at beginning of each log:

![image](https://user-images.githubusercontent.com/5798698/124602982-c0a67980-de69-11eb-829f-7468f0f2b4dc.png)

this is typically happening when the logs are coming from different threads.

A quick look in the `JsonFormatter` showed that there is a field:

      private final StringBuilderWriter writer = new StringBuilderWriter();

Debugger showed that the `JsonFormatter` is a singleton within the our application, so every thread is writing/clearing this single buffer causing this issues?

**Comment by dfa1** [comment]
@gsmet please have a look here :)

PS we just migrated to this library, as per https://github.com/quarkusio/quarkus/issues/15193#issuecomment-782133227

**Comment by SlyngDK** [comment]
@dfa1 Thx for creating the issue, I understand the problem you are pointing at.

I see these different solutions for the problem, with different performance impact.
- `synchronized` like the class the JsonFormatter extends [StructuredFormatter](https://github.com/jboss-logging/jboss-logmanager/blob/08e2781fd26afe43ff1b891348760a273f23d928/ext/src/main/java/org/jboss/logmanager/ext/formatters/StructuredFormatter.java#L210)
- Pool of writers, result in some other locking when requesting at writer not in use.
- Use new writer for each format, results in more object creation(Including auto expand of buffer) and GC.

I don't have any performance numbers of the different solutions, but I think `synchronized` is the best compromise.

@dfa1 If you have the time you could create a micro benchmark, to check the numbers, against current library.

**Comment by dfa1** [comment]
@SlyngDK thank you for the answer :) I would like to have a release if possible, so maybe the `synchronized` is the best option right now (better than my proposed PR).

@SlyngDK later I could try to write a benchmark, also the evaluate the option of pool of writers

**Comment by gsmet** [comment]
@SlyngDK I would just create a new one each time. It's a lightweight object and that's how you use `StringBuilder` usually.

**Comment by dfa1** [comment]
@SlyngDK would it be possible to merge my pull request as is, release it and then iterate on a better fix? @gsmet has a valid argument about StringBuilder)

**Comment by SlyngDK** [comment]
I have done a quick benchmark with 1 and 2 thread, and creating a new StringBuildWriter is the best compromise.

I have released 1.0.1.

```
Benchmark                        Mode  Cnt        Score   Error  Units
BenchmarkNewStringWriter.init1  thrpt    2   598932,618          ops/s
BenchmarkNewStringWriter.init2  thrpt    2  1087659,986          ops/s
BenchmarkObjectPool.init1       thrpt    2   554206,536          ops/s
BenchmarkObjectPool.init2       thrpt    2   837947,587          ops/s
BenchmarkOld.init1              thrpt    2   670263,357          ops/s
BenchmarkOld.init2              thrpt    2  1206957,338          ops/s
BenchmarkSynchronized.init1     thrpt    2   663996,158          ops/s
BenchmarkSynchronized.init2     thrpt    2   577327,380          ops/s
```

**Comment by dfa1** [comment]
thank you @SlyngDK! 

just one question: which object pooling library you used in the benchmark?

**Comment by SlyngDK** [comment]
@dfa1 I used [Apache Commons Pool](https://commons.apache.org/proper/commons-pool/index.html).

**Comment by dfa1** [comment]
@SlyngDK thank you!  Can you please push the benchmark code somewhere? I would like to play a bit with https://github.com/chrisvest/stormpot

**Comment by gsmet** [comment]
You can play for sure but my advice is to keep it simple. `StringBuilder` is not an expensive object to create and it has been designed for exactly this usage.

**Comment by dfa1** [comment]
@gsmet yes, it is just out of curiosity :) By the way the current fix shipped with 1.0.1 is obviously working fine for my use case (few thousands log events per hour)
